### PR TITLE
Introduce --wait flag for karmada unjoin command

### PR DESCRIPTION
**Pull request by wawa0210**
Signed-off-by: wawa0210 <xiaozhang0210@hotmail.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Introduce --wait flag for karmada join and unjoin command
We can control the timeout of join and unjoin by passing the wait parameter, such as 

Two minutes timeout during unjoin
```
karmadactl unjoin CLUSTER_NAME --cluster-kubeconfig=<KUBECONFIG> --wait 2m
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
Nice job, thank you.

/lgtm

**Review by RainbowMango — COMMENTED** [review]
Great thanks to @wawa0210 .
I think the `join` command doesn't need `--wait` flag.

File: pkg/karmadactl/unjoin.go [hunk 1]
```diff
@@ -125,15 +148,15 @@ func RunUnjoin(cmdOut io.Writer, karmadaConfig KarmadaConfig, opts CommandUnjoin
 		}
 	}
 
-	return UnJoinCluster(controlPlaneRestConfig, clusterConfig, opts.ClusterNamespace, opts.ClusterName, opts.forceDeletion, opts.DryRun)
+	return UnJoinCluster(controlPlaneRestConfig, clusterConfig, opts.ClusterNamespace, opts.ClusterName, opts.forceDeletion, opts.DryRun, opts.Wait)
 }
 
 // UnJoinCluster unJoin the cluster from karmada.
-func UnJoinCluster(controlPlaneRestConfig, clusterConfig *rest.Config, clusterNamespace, clusterName string, forceDeletion, dryRun bool) (err error) {
+func UnJoinCluster(controlPlaneRestConfig, clusterConfig *rest.Config, clusterNamespace, clusterName string, forceDeletion, dryRun bool, wait time.Duration) (err error) {
```
[the remark this answers is not in the document]
Comment (wawa0210) [reply pkg/karmadactl/unjoin.go:155]: I have noticed that there are too many parameters in the signature, but in order to ensure the purity of this pr (for review), I did not make adjustments, and a single pr can be used to solve this issue later.

File: pkg/karmadactl/unjoin.go [hunk 2]
```diff
@@ -58,6 +67,10 @@ func NewCmdUnjoin(cmdOut io.Writer, karmadaConfig KarmadaConfig, cmdStr string)
 	return cmd
 }
 
+func getUnjoinExample(cmdStr string) string {
+	return strings.Replace(unjoinExample, "%s", cmdStr, -1)
+}
+
```
[the remark this answers is not in the document]
Comment (wawa0210) [reply pkg/karmadactl/unjoin.go:73]: The number of variables in unjoinExample may change, this time one variable is changed to two variables, so I extracted a separate method to do this

File: pkg/karmadactl/unjoin.go [hunk 3]
```diff
@@ -26,7 +28,11 @@ var (
 	unjoinShort   = `Remove the registration of a cluster from control plane`
 	unjoinLong    = `Unjoin removes the registration of a cluster from control plane.`
 	unjoinExample = `
-%s unjoin CLUSTER_NAME --cluster-kubeconfig=<KUBECONFIG>
+	# Unjoin cluster from karamada control plane
+	%s unjoin CLUSTER_NAME --cluster-kubeconfig=<KUBECONFIG>
+
+	# Unjoin cluster from karamada control plane with timeout
+	%s unjoin CLUSTER_NAME --cluster-kubeconfig=<KUBECONFIG> --wait 2m
```
[the remark this answers is not in the document]
Comment (wawa0210) [reply pkg/karmadactl/unjoin.go]: Thanks, very careful review, update

**Comment by wawa0210** [comment]
> Great thanks to @wawa0210 . I think the `join` command doesn't need `--wait` flag.

I noticed that in the join process, there will be a timeout waiting mechanism for creating sa, so this flag is also added

https://github.com/karmada-io/karmada/blob/e6d7ce369dc5b658ea2383951a0bd3ceb41a6b7f/pkg/karmadactl/join.go#L264

File: pkg/karmadactl/options/global.go [hunk 4]
```diff
@@ -28,4 +38,5 @@ func (o *GlobalCommandOptions) AddFlags(flags *pflag.FlagSet) {
 	flags.StringVar(&o.KarmadaContext, "karmada-context", "", "Name of the cluster context in control plane kubeconfig file.")
 	flags.StringVar(&o.ClusterNamespace, "cluster-namespace", DefaultKarmadaClusterNamespace, "Namespace in the control plane where member cluster are stored.")
 	flags.BoolVar(&o.DryRun, "dry-run", false, "Run the command in dry-run mode, without making any server requests.")
+	flags.DurationVar(&o.Wait, "wait", 60*time.Second, "wait for the command execution process(default 60s), if there is no success after this time, timeout will be returned.")
```
[the remark this answers is not in the document]
Comment (wawa0210) [reply pkg/karmadactl/options/global.go:41]: Previously, it was stated in GlobalCommandOptions that both join and unjoin have wait flags. If join does not require wait, there is no need to put it here.

File: pkg/karmadactl/unjoin.go [hunk 5]
```diff
@@ -90,6 +103,16 @@ func (j *CommandUnjoinOption) Complete(args []string) error {
 	return nil
 }
 
+// Validate ensures that command unjoin options are valid.
+func (j *CommandUnjoinOption) Validate() []error {
+	var errs []error
+
+	if j.Wait < 0 {
+		errs = append(errs, fmt.Errorf(" --wait %v  must be a positive number ", j.Wait))
```
[the remark this answers is not in the document]
Comment (wawa0210) [reply pkg/karmadactl/unjoin.go]: done

File: pkg/karmadactl/unjoin.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/karmadactl/unjoin.go:155]: > I did not make adjustments, and a single pr can be used to solve this issue later.

That makes sense, but can we do it first, then back to this PR?

File: pkg/karmadactl/unjoin.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/karmadactl/unjoin.go:73]: Yeah, more such variables may be added in the future. 
The reason why we add the `%s` is we want to share the code between `kubectl-karmada` and `karmadactl`, it now seems less elegant.

File: pkg/karmadactl/unjoin.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (wawa0210) [reply pkg/karmadactl/unjoin.go:155]: https://github.com/karmada-io/karmada/pull/887

**Pull request merged**